lib/shopify_api/utils/attributes_comparator.rb

```diff
@@ -0,0 +1,66 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "hash_diff"
+
+module ShopifyAPI
+  module Utils
+    module AttributesComparator
+      class << self
+        extend T::Sig
+
+        sig do
+          params(
+            original_attributes: T::Hash[String, T.untyped],
+            updated_attributes: T::Hash[String, T.untyped],
+          ).returns(T::Hash[String, T.untyped])
+        end
+        def compare(original_attributes, updated_attributes)
+          attributes_diff = HashDiff::Comparison.new(
+            original_attributes,
+            updated_attributes,
+          ).left_diff
+
+          update_value = build_update_value(
+            attributes_diff,
+            reference_values: updated_attributes,
+          )
+
+          update_value
+        end
+
+        sig do
+          params(
+            diff: T::Hash[String, T.untyped],
+            path: T::Array[String],
+            reference_values: T::Hash[String, T.untyped],
+          ).returns(T::Hash[String, T.untyped])
+        end
+        def build_update_value(diff, path: [], reference_values: {})
+          new_hash = {}
+
+          diff.each do |key, value|
+            current_path = path + [key.to_s]
+
+            if value.is_a?(Hash)
+              has_numbered_key = value.keys.any? { |k| k.is_a?(Integer) }
+              ref_value = T.unsafe(reference_values).dig(*current_path)
+
+              if has_numbered_key && ref_value.is_a?(Array)
+                new_hash[key] = ref_value
+              else
+                new_value = build_update_value(value, path: current_path, reference_values: reference_values)
+
+                new_hash[key] = new_value unless new_value.empty? && !ref_value.empty?
```

Comment:
in an opposite case, where we only update the address partially but accidentally use the same value for the same attribute, this helps clear out the entire diff because there's nothing to update. It's an edge case. For example

```ruby

original = {
  shipping_address: {
    address1: "foo",
    address2: "bar"
  }
}

updated = {
  shipping_address: {
    address2: "bar"
  }
}

diff = {
  shipping_address: {
    address1: HashDiff::NO_VALUE # hash_diff thinks we want to remove `address1`
  } # address2 doesn't show up because it's not different
}

update_value = {} # should be nothing to update
```